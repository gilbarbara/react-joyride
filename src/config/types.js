import type { Node } from 'react';

export type StateHelpers = {
  close: Function,
  go: Function,
  index: Function,
  info: Function,
  next: Function,
  prev: Function,
  reset: Function,
  restart: Function,
  skip: Function,
  start: Function,
  stop: Function,
};

export type StateInstance = {
  ...StateHelpers,
  update: Function,
};

export type StateObject = {
  action: string,
  controlled: boolean,
  index: number,
  lifecycle: string,
  size: number,
  status: string,
};

type placement = 'top' | 'top-start' | 'top-end' |
  'bottom' | 'bottom-start' | 'bottom-end' |
  'left' | 'left-start' | 'left-end' |
  'right' | 'right-start' | 'right-end' |
  'auto' | 'center';

type placementBeacon = 'top' | 'bottom' | 'left' | 'right';

export type StepProps = {
  beaconComponent: ?Node,
  content: Node | string,
  disableBeacon: boolean,
  disableCloseOnEsc: boolean,
  disableOverlay: boolean,
  disableOverlayClose: boolean,
  disableScrolling: boolean,
  event: string,
  floaterProps: ?Object,
  hideBackButton: ?boolean,
  isFixed: ?boolean,
  offset: number,
  placement: placement,
  placementBeacon: placementBeacon,
  showProgress: ?boolean,
  showSkipButton: ?boolean,
  spotlightPadding: number,
  spotlightClicks: boolean,
  styles: ?Object,
  target: string | HTMLElement,
  title: ?Node,
  tooltipComponent: ?Node,
}

export type JoyrideProps = {
  beaconComponent: ?Node,
  callback: ?Function,
  continuous: boolean,
  debug: boolean,
  disableCloseOnEsc: boolean,
  disableOverlay: boolean,
  disableOverlayClose: boolean,
  disableScrolling: boolean,
  floaterProps: ?Object,
  hideBackButton: boolean,
  locale: ?Object,
  run: boolean,
  scrollOffset: number,
  scrollToFirstStep: boolean,
  showProgress: boolean,
  showSkipButton: boolean,
  spotlightPadding: boolean,
  spotlightClicks: boolean,
  stepIndex: ?number,
  steps: Array<StepProps>,
  styles: ?Object,
  tooltipComponent: ?Node,
}

export type CallBackProps = {
  action: string,
  controlled: boolean,
  index: number,
  lifecycle: string,
  size: number,
  status: string,
  step: StepProps,
  type: string,
}
