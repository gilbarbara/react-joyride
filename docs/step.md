# Step

The step is an plain object that only requires two properties to be valid: `target` and `content` \(or `tooltipComponent`\).

```text
{
  target: '.my-selector',
  content: 'This is my super awesome feature!'
}
```

It will inherit some properties from the Joyride's own [props](props.md) that can be overridden per step:

* beaconComponent
* disableCloseOnEsc
* disableOverlay
* disableOverlayClose
* disableScrolling
* floaterProps (check the [getMergedStep](../src/modules/step.js) function for more information)
* hideBackButton
* locale
* showProgress
* showSkipButton
* spotlightClicks
* spotlightPadding
* styles
* tooltipComponent

## And you can use these

**content** {React.Node\|string}  
The tooltip's body.

**disableBeacon** {boolean} ▶︎ `false`  
Don't show the Beacon before the tooltip.

**event** {string} ▶︎ `click`  
The event to trigger the beacon. It can be _**click**_ or _**hover**_

**isFixed** {boolean} ▶︎ `false`  
Force the step to be fixed.

**offset** {React.Node\|string} ▶︎ `10`  
The distance from the target to the tooltip.

**placement** {string} ▶︎ `bottom`  
The placement of the beacon and tooltip. It will re-position itself if there's no space available.  
It can be:

* top \(top-start, top-end\)
* bottom \(bottom-start, bottom-end\)
* left \(left-start, left-end\)
* right \(right-start, right-end
* auto
* center

Check [react-floater](https://github.com/gilbarbara/react-floater) for more information.

**placementBeacon** {string} ▶︎ placement  
The placement of the beacon. It will use the placement if nothing is passed and it can be: `top, bottom, left, right`.

**styles** {Object}  
Override the [styling](styling.md) of the step's Tooltip

**target** {Element\|string} - **required**  
The target for the step. It can be a [CSS selector](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Selectors) or an HtmlElement directly \(but using refs created in the same render would required an additional render afterwards\).

**title** {React.Node\|string}  
The tooltip's title.

