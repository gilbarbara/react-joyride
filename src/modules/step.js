// @flow
import deepmerge from 'deepmerge';
import is from 'is-lite';

import { getElement, hasCustomScrollParent } from './dom';
import { log } from './helpers';
import getStyles from '../styles';

import DEFAULTS from '../config/defaults';

import type { StepProps, JoyrideProps } from '../config/types';

/**
 * Validate if a step is valid
 *
 * @param {Object} step - A step object
 * @param {boolean} debug
 *
 * @returns {boolean} - True if the step is valid, false otherwise
 */
export function validateStep(step: StepProps, debug: boolean = false): boolean {
  if (!is.plainObject(step)) {
    log({
      title: 'validateStep',
      data: 'step must be an object',
      warn: true,
      debug,
    });
    return false;
  }

  if (!step.target) {
    log({
      title: 'validateStep',
      data: 'target is missing from the step',
      warn: true,
      debug,
    });
    return false;
  }

  return true;
}

/**
 * Validate if steps is valid
 *
 * @param {Array} steps - A steps array
 * @param {boolean} debug
 *
 * @returns {boolean} - True if the steps are valid, false otherwise
 */
export function validateSteps(steps: Array<Object>, debug: boolean = false): boolean {
  if (!is.array(steps)) {
    log({
      title: 'validateSteps',
      data: 'steps must be an array',
      warn: true,
      debug,
    });

    return false;
  }

  return steps.every(d => validateStep(d, debug));
}

function getTourProps(props: JoyrideProps): JoyrideProps {
  const sharedTourProps = [
    'beaconComponent',
    'disableCloseOnEsc',
    'disableOverlay',
    'disableOverlayClose',
    'disableScrolling',
    'floaterProps',
    'hideBackButton',
    'locale',
    'showProgress',
    'showSkipButton',
    'spotlightClicks',
    'spotlightPadding',
    'styles',
    'tooltipComponent',
  ];

  return Object.keys(props)
    .filter(d => sharedTourProps.includes(d))
    .reduce((acc, i) => {
      acc[i] = props[i]; //eslint-disable-line react/destructuring-assignment

      return acc;
    }, {});
}

export function getMergedStep(step: StepProps, props: JoyrideProps): StepProps {
  if (!step) return undefined;

  const mergedStep = deepmerge.all([getTourProps(props), DEFAULTS.step, step]);
  const mergedStyles = getStyles(deepmerge(props.styles || {}, step.styles || {}));
  const scrollParent = hasCustomScrollParent(getElement(step.target));
  const floaterProps = deepmerge.all([props.floaterProps || {}, DEFAULTS.floaterProps, mergedStep.floaterProps || {}]);

  // Set react-floater props
  floaterProps.offset = mergedStep.offset;
  floaterProps.styles = deepmerge(floaterProps.styles || {}, mergedStyles.floaterStyles || {});

  delete mergedStyles.floaterStyles;

  if (mergedStep.floaterProps && mergedStep.floaterProps.offset) {
    floaterProps.offset = mergedStep.floaterProps.offset;
  }

  if (!mergedStep.disableScrolling) {
    floaterProps.offset += props.spotlightPadding || step.spotlightPadding || 0;
  }

  if (step.placementBeacon) {
    floaterProps.wrapperOptions.placement = step.placementBeacon;
  }

  if (scrollParent) {
    floaterProps.options.preventOverflow.boundariesElement = 'window';
  }

  return {
    ...mergedStep,
    locale: deepmerge(DEFAULTS.locale, props.locale || {}),
    floaterProps,
    styles: mergedStyles,
  };
}
