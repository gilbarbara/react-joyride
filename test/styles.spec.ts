import { getMergedStep } from '~/modules/step';

import getStyles from '~/styles';
import { Props, Step } from '~/types';

const props: Props = {
  run: false,
  steps: [],
};

const customStylesProps = {
  ...props,
  styles: {
    buttonNext: {
      color: '#00f',
    },
    options: {
      arrowColor: '#00f',
      primaryColor: '#ff0',
      backgroundColor: '#00f',
      textColor: '#fff',
    },
  },
};

const baseStep: Step = {
  content: 'Hello',
  target: 'body',
};

describe('styles', () => {
  it('should return the default styles', () => {
    expect(getStyles(props, getMergedStep(props, baseStep))).toMatchSnapshot();
  });

  it("should return the component floaterProps' arrow color instead of styles.options.arrowColor", () => {
    expect(
      getStyles(
        {
          ...props,
          floaterProps: { styles: { arrow: { color: '#f00' } } },
          styles: { options: { arrowColor: '#00f' } },
        },
        getMergedStep(props, baseStep),
      ),
    ).toMatchSnapshot();
  });

  it("should return the step floaterProps' arrow color instead of styles.options.arrowColor", () => {
    expect(
      getStyles(
        props,
        getMergedStep(props, {
          ...baseStep,
          floaterProps: { styles: { arrow: { color: '#f00' } } },
          styles: { options: { arrowColor: '#00f' } },
        }),
      ),
    ).toMatchSnapshot();
  });

  it("should return the inherited props' styles", () => {
    expect(
      getStyles(customStylesProps, getMergedStep(customStylesProps, baseStep)),
    ).toMatchSnapshot();
  });

  it("should return the step's styles", () => {
    const step = {
      ...baseStep,
      styles: {
        buttonNext: {
          color: '#000',
        },
        options: {
          arrowColor: '#000',
          primaryColor: '#f00',
          backgroundColor: '#000',
          textColor: '#fff',
        },
      },
    };

    expect(getStyles(customStylesProps, getMergedStep(customStylesProps, step))).toMatchSnapshot();
  });
});
