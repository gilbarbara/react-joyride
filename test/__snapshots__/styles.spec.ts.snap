// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`styles should return the component floaterProps' arrow color instead of styles.options.arrowColor 1`] = `
{
  "beacon": {
    "WebkitAppearance": "none",
    "backgroundColor": "transparent",
    "border": 0,
    "borderRadius": 0,
    "color": "#555",
    "cursor": "pointer",
    "display": "inline-block",
    "fontSize": 16,
    "height": 36,
    "lineHeight": 1,
    "padding": 8,
    "position": "relative",
    "width": 36,
    "zIndex": 100,
  },
  "beaconInner": {
    "animation": "joyride-beacon-inner 1.2s infinite ease-in-out",
    "backgroundColor": "#f04",
    "borderRadius": "50%",
    "display": "block",
    "height": "50%",
    "left": "50%",
    "opacity": 0.7,
    "position": "absolute",
    "top": "50%",
    "transform": "translate(-50%, -50%)",
    "width": "50%",
  },
  "beaconOuter": {
    "animation": "joyride-beacon-outer 1.2s infinite ease-in-out",
    "backgroundColor": "rgba(255,0,68, 0.2)",
    "border": "2px solid #f04",
    "borderRadius": "50%",
    "boxSizing": "border-box",
    "display": "block",
    "height": "100%",
    "left": 0,
    "opacity": 0.9,
    "position": "absolute",
    "top": 0,
    "transformOrigin": "center",
    "width": "100%",
  },
  "buttonBack": {
    "WebkitAppearance": "none",
    "backgroundColor": "transparent",
    "border": 0,
    "borderRadius": 0,
    "color": "#f04",
    "cursor": "pointer",
    "fontSize": 16,
    "lineHeight": 1,
    "marginLeft": "auto",
    "marginRight": 5,
    "padding": 8,
  },
  "buttonClose": {
    "WebkitAppearance": "none",
    "backgroundColor": "transparent",
    "border": 0,
    "borderRadius": 0,
    "color": "#333",
    "cursor": "pointer",
    "fontSize": 16,
    "height": 14,
    "lineHeight": 1,
    "padding": 15,
    "position": "absolute",
    "right": 0,
    "top": 0,
    "width": 14,
  },
  "buttonNext": {
    "WebkitAppearance": "none",
    "backgroundColor": "#f04",
    "border": 0,
    "borderRadius": 4,
    "color": "#fff",
    "cursor": "pointer",
    "fontSize": 16,
    "lineHeight": 1,
    "padding": 8,
  },
  "buttonSkip": {
    "WebkitAppearance": "none",
    "backgroundColor": "transparent",
    "border": 0,
    "borderRadius": 0,
    "color": "#333",
    "cursor": "pointer",
    "fontSize": 14,
    "lineHeight": 1,
    "padding": 8,
  },
  "floaterStyles": {
    "arrow": {
      "color": "#f00",
    },
    "options": {
      "zIndex": 200,
    },
  },
  "options": {
    "arrowColor": "#fff",
    "backgroundColor": "#fff",
    "beaconSize": 36,
    "overlayColor": "rgba(0, 0, 0, 0.5)",
    "primaryColor": "#f04",
    "spotlightShadow": "0 0 15px rgba(0, 0, 0, 0.5)",
    "textColor": "#333",
    "width": 380,
    "zIndex": 100,
  },
  "overlay": {
    "backgroundColor": "rgba(0, 0, 0, 0.5)",
    "bottom": 0,
    "left": 0,
    "mixBlendMode": "hard-light",
    "overflow": "hidden",
    "position": "absolute",
    "right": 0,
    "top": 0,
    "zIndex": 100,
  },
  "overlayLegacy": {
    "bottom": 0,
    "left": 0,
    "overflow": "hidden",
    "position": "absolute",
    "right": 0,
    "top": 0,
    "zIndex": 100,
  },
  "overlayLegacyCenter": {
    "backgroundColor": "rgba(0, 0, 0, 0.5)",
    "bottom": 0,
    "left": 0,
    "overflow": "hidden",
    "position": "absolute",
    "right": 0,
    "top": 0,
    "zIndex": 100,
  },
  "spotlight": {
    "backgroundColor": "gray",
    "borderRadius": 4,
    "position": "absolute",
  },
  "spotlightLegacy": {
    "borderRadius": 4,
    "boxShadow": "0 0 0 9999px rgba(0, 0, 0, 0.5), 0 0 15px rgba(0, 0, 0, 0.5)",
    "position": "absolute",
  },
  "tooltip": {
    "backgroundColor": "#fff",
    "borderRadius": 5,
    "boxSizing": "border-box",
    "color": "#333",
    "fontSize": 16,
    "maxWidth": "100%",
    "padding": 15,
    "position": "relative",
    "width": 380,
  },
  "tooltipContainer": {
    "lineHeight": 1.4,
    "textAlign": "center",
  },
  "tooltipContent": {
    "padding": "20px 10px",
  },
  "tooltipFooter": {
    "alignItems": "center",
    "display": "flex",
    "justifyContent": "flex-end",
    "marginTop": 15,
  },
  "tooltipFooterSpacer": {
    "flex": 1,
  },
  "tooltipTitle": {
    "fontSize": 18,
    "margin": 0,
  },
}
`;

exports[`styles should return the default styles 1`] = `
{
  "beacon": {
    "WebkitAppearance": "none",
    "backgroundColor": "transparent",
    "border": 0,
    "borderRadius": 0,
    "color": "#555",
    "cursor": "pointer",
    "display": "inline-block",
    "fontSize": 16,
    "height": 36,
    "lineHeight": 1,
    "padding": 8,
    "position": "relative",
    "width": 36,
    "zIndex": 100,
  },
  "beaconInner": {
    "animation": "joyride-beacon-inner 1.2s infinite ease-in-out",
    "backgroundColor": "#f04",
    "borderRadius": "50%",
    "display": "block",
    "height": "50%",
    "left": "50%",
    "opacity": 0.7,
    "position": "absolute",
    "top": "50%",
    "transform": "translate(-50%, -50%)",
    "width": "50%",
  },
  "beaconOuter": {
    "animation": "joyride-beacon-outer 1.2s infinite ease-in-out",
    "backgroundColor": "rgba(255,0,68, 0.2)",
    "border": "2px solid #f04",
    "borderRadius": "50%",
    "boxSizing": "border-box",
    "display": "block",
    "height": "100%",
    "left": 0,
    "opacity": 0.9,
    "position": "absolute",
    "top": 0,
    "transformOrigin": "center",
    "width": "100%",
  },
  "buttonBack": {
    "WebkitAppearance": "none",
    "backgroundColor": "transparent",
    "border": 0,
    "borderRadius": 0,
    "color": "#f04",
    "cursor": "pointer",
    "fontSize": 16,
    "lineHeight": 1,
    "marginLeft": "auto",
    "marginRight": 5,
    "padding": 8,
  },
  "buttonClose": {
    "WebkitAppearance": "none",
    "backgroundColor": "transparent",
    "border": 0,
    "borderRadius": 0,
    "color": "#333",
    "cursor": "pointer",
    "fontSize": 16,
    "height": 14,
    "lineHeight": 1,
    "padding": 15,
    "position": "absolute",
    "right": 0,
    "top": 0,
    "width": 14,
  },
  "buttonNext": {
    "WebkitAppearance": "none",
    "backgroundColor": "#f04",
    "border": 0,
    "borderRadius": 4,
    "color": "#fff",
    "cursor": "pointer",
    "fontSize": 16,
    "lineHeight": 1,
    "padding": 8,
  },
  "buttonSkip": {
    "WebkitAppearance": "none",
    "backgroundColor": "transparent",
    "border": 0,
    "borderRadius": 0,
    "color": "#333",
    "cursor": "pointer",
    "fontSize": 14,
    "lineHeight": 1,
    "padding": 8,
  },
  "floaterStyles": {
    "arrow": {
      "color": "#fff",
    },
    "options": {
      "zIndex": 200,
    },
  },
  "options": {
    "arrowColor": "#fff",
    "backgroundColor": "#fff",
    "beaconSize": 36,
    "overlayColor": "rgba(0, 0, 0, 0.5)",
    "primaryColor": "#f04",
    "spotlightShadow": "0 0 15px rgba(0, 0, 0, 0.5)",
    "textColor": "#333",
    "width": 380,
    "zIndex": 100,
  },
  "overlay": {
    "backgroundColor": "rgba(0, 0, 0, 0.5)",
    "bottom": 0,
    "left": 0,
    "mixBlendMode": "hard-light",
    "overflow": "hidden",
    "position": "absolute",
    "right": 0,
    "top": 0,
    "zIndex": 100,
  },
  "overlayLegacy": {
    "bottom": 0,
    "left": 0,
    "overflow": "hidden",
    "position": "absolute",
    "right": 0,
    "top": 0,
    "zIndex": 100,
  },
  "overlayLegacyCenter": {
    "backgroundColor": "rgba(0, 0, 0, 0.5)",
    "bottom": 0,
    "left": 0,
    "overflow": "hidden",
    "position": "absolute",
    "right": 0,
    "top": 0,
    "zIndex": 100,
  },
  "spotlight": {
    "backgroundColor": "gray",
    "borderRadius": 4,
    "position": "absolute",
  },
  "spotlightLegacy": {
    "borderRadius": 4,
    "boxShadow": "0 0 0 9999px rgba(0, 0, 0, 0.5), 0 0 15px rgba(0, 0, 0, 0.5)",
    "position": "absolute",
  },
  "tooltip": {
    "backgroundColor": "#fff",
    "borderRadius": 5,
    "boxSizing": "border-box",
    "color": "#333",
    "fontSize": 16,
    "maxWidth": "100%",
    "padding": 15,
    "position": "relative",
    "width": 380,
  },
  "tooltipContainer": {
    "lineHeight": 1.4,
    "textAlign": "center",
  },
  "tooltipContent": {
    "padding": "20px 10px",
  },
  "tooltipFooter": {
    "alignItems": "center",
    "display": "flex",
    "justifyContent": "flex-end",
    "marginTop": 15,
  },
  "tooltipFooterSpacer": {
    "flex": 1,
  },
  "tooltipTitle": {
    "fontSize": 18,
    "margin": 0,
  },
}
`;

exports[`styles should return the inherited props' styles 1`] = `
{
  "beacon": {
    "WebkitAppearance": "none",
    "backgroundColor": "transparent",
    "border": 0,
    "borderRadius": 0,
    "color": "#555",
    "cursor": "pointer",
    "display": "inline-block",
    "fontSize": 16,
    "height": 36,
    "lineHeight": 1,
    "padding": 8,
    "position": "relative",
    "width": 36,
    "zIndex": 100,
  },
  "beaconInner": {
    "animation": "joyride-beacon-inner 1.2s infinite ease-in-out",
    "backgroundColor": "#ff0",
    "borderRadius": "50%",
    "display": "block",
    "height": "50%",
    "left": "50%",
    "opacity": 0.7,
    "position": "absolute",
    "top": "50%",
    "transform": "translate(-50%, -50%)",
    "width": "50%",
  },
  "beaconOuter": {
    "animation": "joyride-beacon-outer 1.2s infinite ease-in-out",
    "backgroundColor": "rgba(255,255,0, 0.2)",
    "border": "2px solid #ff0",
    "borderRadius": "50%",
    "boxSizing": "border-box",
    "display": "block",
    "height": "100%",
    "left": 0,
    "opacity": 0.9,
    "position": "absolute",
    "top": 0,
    "transformOrigin": "center",
    "width": "100%",
  },
  "buttonBack": {
    "WebkitAppearance": "none",
    "backgroundColor": "transparent",
    "border": 0,
    "borderRadius": 0,
    "color": "#ff0",
    "cursor": "pointer",
    "fontSize": 16,
    "lineHeight": 1,
    "marginLeft": "auto",
    "marginRight": 5,
    "padding": 8,
  },
  "buttonClose": {
    "WebkitAppearance": "none",
    "backgroundColor": "transparent",
    "border": 0,
    "borderRadius": 0,
    "color": "#fff",
    "cursor": "pointer",
    "fontSize": 16,
    "height": 14,
    "lineHeight": 1,
    "padding": 15,
    "position": "absolute",
    "right": 0,
    "top": 0,
    "width": 14,
  },
  "buttonNext": {
    "WebkitAppearance": "none",
    "backgroundColor": "#ff0",
    "border": 0,
    "borderRadius": 4,
    "color": "#00f",
    "cursor": "pointer",
    "fontSize": 16,
    "lineHeight": 1,
    "padding": 8,
  },
  "buttonSkip": {
    "WebkitAppearance": "none",
    "backgroundColor": "transparent",
    "border": 0,
    "borderRadius": 0,
    "color": "#fff",
    "cursor": "pointer",
    "fontSize": 14,
    "lineHeight": 1,
    "padding": 8,
  },
  "floaterStyles": {
    "arrow": {
      "color": "#00f",
    },
    "options": {
      "zIndex": 200,
    },
  },
  "options": {
    "arrowColor": "#00f",
    "backgroundColor": "#00f",
    "beaconSize": 36,
    "overlayColor": "rgba(0, 0, 0, 0.5)",
    "primaryColor": "#ff0",
    "spotlightShadow": "0 0 15px rgba(0, 0, 0, 0.5)",
    "textColor": "#fff",
    "width": 380,
    "zIndex": 100,
  },
  "overlay": {
    "backgroundColor": "rgba(0, 0, 0, 0.5)",
    "bottom": 0,
    "left": 0,
    "mixBlendMode": "hard-light",
    "overflow": "hidden",
    "position": "absolute",
    "right": 0,
    "top": 0,
    "zIndex": 100,
  },
  "overlayLegacy": {
    "bottom": 0,
    "left": 0,
    "overflow": "hidden",
    "position": "absolute",
    "right": 0,
    "top": 0,
    "zIndex": 100,
  },
  "overlayLegacyCenter": {
    "backgroundColor": "rgba(0, 0, 0, 0.5)",
    "bottom": 0,
    "left": 0,
    "overflow": "hidden",
    "position": "absolute",
    "right": 0,
    "top": 0,
    "zIndex": 100,
  },
  "spotlight": {
    "backgroundColor": "gray",
    "borderRadius": 4,
    "position": "absolute",
  },
  "spotlightLegacy": {
    "borderRadius": 4,
    "boxShadow": "0 0 0 9999px rgba(0, 0, 0, 0.5), 0 0 15px rgba(0, 0, 0, 0.5)",
    "position": "absolute",
  },
  "tooltip": {
    "backgroundColor": "#00f",
    "borderRadius": 5,
    "boxSizing": "border-box",
    "color": "#fff",
    "fontSize": 16,
    "maxWidth": "100%",
    "padding": 15,
    "position": "relative",
    "width": 380,
  },
  "tooltipContainer": {
    "lineHeight": 1.4,
    "textAlign": "center",
  },
  "tooltipContent": {
    "padding": "20px 10px",
  },
  "tooltipFooter": {
    "alignItems": "center",
    "display": "flex",
    "justifyContent": "flex-end",
    "marginTop": 15,
  },
  "tooltipFooterSpacer": {
    "flex": 1,
  },
  "tooltipTitle": {
    "fontSize": 18,
    "margin": 0,
  },
}
`;

exports[`styles should return the step floaterProps' arrow color instead of styles.options.arrowColor 1`] = `
{
  "beacon": {
    "WebkitAppearance": "none",
    "backgroundColor": "transparent",
    "border": 0,
    "borderRadius": 0,
    "color": "#555",
    "cursor": "pointer",
    "display": "inline-block",
    "fontSize": 16,
    "height": 36,
    "lineHeight": 1,
    "padding": 8,
    "position": "relative",
    "width": 36,
    "zIndex": 100,
  },
  "beaconInner": {
    "animation": "joyride-beacon-inner 1.2s infinite ease-in-out",
    "backgroundColor": "#f04",
    "borderRadius": "50%",
    "display": "block",
    "height": "50%",
    "left": "50%",
    "opacity": 0.7,
    "position": "absolute",
    "top": "50%",
    "transform": "translate(-50%, -50%)",
    "width": "50%",
  },
  "beaconOuter": {
    "animation": "joyride-beacon-outer 1.2s infinite ease-in-out",
    "backgroundColor": "rgba(255,0,68, 0.2)",
    "border": "2px solid #f04",
    "borderRadius": "50%",
    "boxSizing": "border-box",
    "display": "block",
    "height": "100%",
    "left": 0,
    "opacity": 0.9,
    "position": "absolute",
    "top": 0,
    "transformOrigin": "center",
    "width": "100%",
  },
  "buttonBack": {
    "WebkitAppearance": "none",
    "backgroundColor": "transparent",
    "border": 0,
    "borderRadius": 0,
    "color": "#f04",
    "cursor": "pointer",
    "fontSize": 16,
    "lineHeight": 1,
    "marginLeft": "auto",
    "marginRight": 5,
    "padding": 8,
  },
  "buttonClose": {
    "WebkitAppearance": "none",
    "backgroundColor": "transparent",
    "border": 0,
    "borderRadius": 0,
    "color": "#333",
    "cursor": "pointer",
    "fontSize": 16,
    "height": 14,
    "lineHeight": 1,
    "padding": 15,
    "position": "absolute",
    "right": 0,
    "top": 0,
    "width": 14,
  },
  "buttonNext": {
    "WebkitAppearance": "none",
    "backgroundColor": "#f04",
    "border": 0,
    "borderRadius": 4,
    "color": "#fff",
    "cursor": "pointer",
    "fontSize": 16,
    "lineHeight": 1,
    "padding": 8,
  },
  "buttonSkip": {
    "WebkitAppearance": "none",
    "backgroundColor": "transparent",
    "border": 0,
    "borderRadius": 0,
    "color": "#333",
    "cursor": "pointer",
    "fontSize": 14,
    "lineHeight": 1,
    "padding": 8,
  },
  "floaterStyles": {
    "arrow": {
      "color": "#f00",
    },
    "options": {
      "zIndex": 200,
    },
  },
  "options": {
    "arrowColor": "#00f",
    "backgroundColor": "#fff",
    "beaconSize": 36,
    "overlayColor": "rgba(0, 0, 0, 0.5)",
    "primaryColor": "#f04",
    "spotlightShadow": "0 0 15px rgba(0, 0, 0, 0.5)",
    "textColor": "#333",
    "width": 380,
    "zIndex": 100,
  },
  "overlay": {
    "backgroundColor": "rgba(0, 0, 0, 0.5)",
    "bottom": 0,
    "left": 0,
    "mixBlendMode": "hard-light",
    "overflow": "hidden",
    "position": "absolute",
    "right": 0,
    "top": 0,
    "zIndex": 100,
  },
  "overlayLegacy": {
    "bottom": 0,
    "left": 0,
    "overflow": "hidden",
    "position": "absolute",
    "right": 0,
    "top": 0,
    "zIndex": 100,
  },
  "overlayLegacyCenter": {
    "backgroundColor": "rgba(0, 0, 0, 0.5)",
    "bottom": 0,
    "left": 0,
    "overflow": "hidden",
    "position": "absolute",
    "right": 0,
    "top": 0,
    "zIndex": 100,
  },
  "spotlight": {
    "backgroundColor": "gray",
    "borderRadius": 4,
    "position": "absolute",
  },
  "spotlightLegacy": {
    "borderRadius": 4,
    "boxShadow": "0 0 0 9999px rgba(0, 0, 0, 0.5), 0 0 15px rgba(0, 0, 0, 0.5)",
    "position": "absolute",
  },
  "tooltip": {
    "backgroundColor": "#fff",
    "borderRadius": 5,
    "boxSizing": "border-box",
    "color": "#333",
    "fontSize": 16,
    "maxWidth": "100%",
    "padding": 15,
    "position": "relative",
    "width": 380,
  },
  "tooltipContainer": {
    "lineHeight": 1.4,
    "textAlign": "center",
  },
  "tooltipContent": {
    "padding": "20px 10px",
  },
  "tooltipFooter": {
    "alignItems": "center",
    "display": "flex",
    "justifyContent": "flex-end",
    "marginTop": 15,
  },
  "tooltipFooterSpacer": {
    "flex": 1,
  },
  "tooltipTitle": {
    "fontSize": 18,
    "margin": 0,
  },
}
`;

exports[`styles should return the step's styles 1`] = `
{
  "beacon": {
    "WebkitAppearance": "none",
    "backgroundColor": "transparent",
    "border": 0,
    "borderRadius": 0,
    "color": "#555",
    "cursor": "pointer",
    "display": "inline-block",
    "fontSize": 16,
    "height": 36,
    "lineHeight": 1,
    "padding": 8,
    "position": "relative",
    "width": 36,
    "zIndex": 100,
  },
  "beaconInner": {
    "animation": "joyride-beacon-inner 1.2s infinite ease-in-out",
    "backgroundColor": "#f00",
    "borderRadius": "50%",
    "display": "block",
    "height": "50%",
    "left": "50%",
    "opacity": 0.7,
    "position": "absolute",
    "top": "50%",
    "transform": "translate(-50%, -50%)",
    "width": "50%",
  },
  "beaconOuter": {
    "animation": "joyride-beacon-outer 1.2s infinite ease-in-out",
    "backgroundColor": "rgba(255,0,0, 0.2)",
    "border": "2px solid #f00",
    "borderRadius": "50%",
    "boxSizing": "border-box",
    "display": "block",
    "height": "100%",
    "left": 0,
    "opacity": 0.9,
    "position": "absolute",
    "top": 0,
    "transformOrigin": "center",
    "width": "100%",
  },
  "buttonBack": {
    "WebkitAppearance": "none",
    "backgroundColor": "transparent",
    "border": 0,
    "borderRadius": 0,
    "color": "#f00",
    "cursor": "pointer",
    "fontSize": 16,
    "lineHeight": 1,
    "marginLeft": "auto",
    "marginRight": 5,
    "padding": 8,
  },
  "buttonClose": {
    "WebkitAppearance": "none",
    "backgroundColor": "transparent",
    "border": 0,
    "borderRadius": 0,
    "color": "#fff",
    "cursor": "pointer",
    "fontSize": 16,
    "height": 14,
    "lineHeight": 1,
    "padding": 15,
    "position": "absolute",
    "right": 0,
    "top": 0,
    "width": 14,
  },
  "buttonNext": {
    "WebkitAppearance": "none",
    "backgroundColor": "#f00",
    "border": 0,
    "borderRadius": 4,
    "color": "#000",
    "cursor": "pointer",
    "fontSize": 16,
    "lineHeight": 1,
    "padding": 8,
  },
  "buttonSkip": {
    "WebkitAppearance": "none",
    "backgroundColor": "transparent",
    "border": 0,
    "borderRadius": 0,
    "color": "#fff",
    "cursor": "pointer",
    "fontSize": 14,
    "lineHeight": 1,
    "padding": 8,
  },
  "floaterStyles": {
    "arrow": {
      "color": "#000",
    },
    "options": {
      "zIndex": 200,
    },
  },
  "options": {
    "arrowColor": "#000",
    "backgroundColor": "#000",
    "beaconSize": 36,
    "overlayColor": "rgba(0, 0, 0, 0.5)",
    "primaryColor": "#f00",
    "spotlightShadow": "0 0 15px rgba(0, 0, 0, 0.5)",
    "textColor": "#fff",
    "width": 380,
    "zIndex": 100,
  },
  "overlay": {
    "backgroundColor": "rgba(0, 0, 0, 0.5)",
    "bottom": 0,
    "left": 0,
    "mixBlendMode": "hard-light",
    "overflow": "hidden",
    "position": "absolute",
    "right": 0,
    "top": 0,
    "zIndex": 100,
  },
  "overlayLegacy": {
    "bottom": 0,
    "left": 0,
    "overflow": "hidden",
    "position": "absolute",
    "right": 0,
    "top": 0,
    "zIndex": 100,
  },
  "overlayLegacyCenter": {
    "backgroundColor": "rgba(0, 0, 0, 0.5)",
    "bottom": 0,
    "left": 0,
    "overflow": "hidden",
    "position": "absolute",
    "right": 0,
    "top": 0,
    "zIndex": 100,
  },
  "spotlight": {
    "backgroundColor": "gray",
    "borderRadius": 4,
    "position": "absolute",
  },
  "spotlightLegacy": {
    "borderRadius": 4,
    "boxShadow": "0 0 0 9999px rgba(0, 0, 0, 0.5), 0 0 15px rgba(0, 0, 0, 0.5)",
    "position": "absolute",
  },
  "tooltip": {
    "backgroundColor": "#000",
    "borderRadius": 5,
    "boxSizing": "border-box",
    "color": "#fff",
    "fontSize": 16,
    "maxWidth": "100%",
    "padding": 15,
    "position": "relative",
    "width": 380,
  },
  "tooltipContainer": {
    "lineHeight": 1.4,
    "textAlign": "center",
  },
  "tooltipContent": {
    "padding": "20px 10px",
  },
  "tooltipFooter": {
    "alignItems": "center",
    "display": "flex",
    "justifyContent": "flex-end",
    "marginTop": 15,
  },
  "tooltipFooterSpacer": {
    "flex": 1,
  },
  "tooltipTitle": {
    "fontSize": 18,
    "margin": 0,
  },
}
`;
